pkg/plugins/manager.go

```diff
@@ -11,160 +12,104 @@ import (
+type PluginAuth struct {
+	Key   string
+	Value string
 }
-*/
-type PluginTypeNameMap map[PluginType]pluginTypeConfigMap
 
-// The Manager holds the different types of plugins that can be used in the plugin system
-type Manager struct {
-	pluginTypes PluginTypeNameMap
+// The Manager holds a map for different kinds of possible plugin clients.
+// For example, the clients map might contain a mapping of "delta" -> plugin.Client to communicate with the Delta
+// plugin.
+type Manager[T any] struct {
+	clients map[string]*plugin.Client
 }
 
-// WrapPlugin generates a Wrapper that wraps the go-plugin client.
-//
-// It accepts two parameters: the top PluginType and the required PluginName under it.
-// For example, PluginType = "diff", PluginName = "delta" will generate a Wrapper with a client that performs
-// diffs over Delta Lake tables.
-// It also returns a DestroyClientFunc
-func (m *Manager) WrapPlugin(pluginType PluginType, pluginName PluginName) (*Wrapper, error) {
-	ptpp, ok := m.pluginTypes[pluginType]
-	if !ok {
-		return nil, errPluginGroupNotFound
+func NewManager[T any]() *Manager[T] {
+	return &Manager[T]{
+		clients: make(map[string]*plugin.Client),
 	}
-	clientConfig, ok := ptpp[pluginName]
+}
+
+// RegisterPlugin is used to register a new plugin client with the corresponding plugin type.
+func (m *Manager[T]) RegisterPlugin(name string, id PluginIdentity, auth PluginAuth, p plugin.Plugin) error {
+	if m == nil {
+		return ErrUninitializedManager
+	}
+	hc := plugin.HandshakeConfig{
+		ProtocolVersion:  uint(id.Version),
+		MagicCookieKey:   auth.Key,
+		MagicCookieValue: auth.Value,
+	}
+	cmd := exec.Command(id.ExecutableLocation) //nolint:gosec
+	c, err := pluginClient(name, p, hc, cmd)
+	if err != nil {
+		return err
+	}
+	_, ok := any(c).(T)
 	if !ok {
-		return nil, errPluginTypeNotFound
+		return ErrInvalidPluginNotFound
 	}
-	return newPluginWrapper(fmt.Sprintf("%s_%s", pluginType, pluginName), clientConfig)
+	m.clients[name] = c
+	return nil
 }
 
-// newPluginWrapper generates a Wrapper that wraps the go-plugin client.
-func newPluginWrapper(clientName string, clientConfig plugin.ClientConfig) (*Wrapper, error) {
-	logger := hclog.New(&hclog.LoggerOptions{
+func pluginClient(name string, p plugin.Plugin, hc plugin.HandshakeConfig, cmd *exec.Cmd) (*plugin.Client, error) {
+	clientConfig := plugin.ClientConfig{
+		Plugins: map[string]plugin.Plugin{
+			name: p,
+		},
+		AllowedProtocols: allowedProtocols,
+		HandshakeConfig:  hc,
+		Cmd:              cmd,
+	}
+	return newPluginClient(name, clientConfig)
+}
+
+func newPluginClient(clientName string, clientConfig plugin.ClientConfig) (*plugin.Client, error) {
+	hl := hclog.New(&hclog.LoggerOptions{
 		Name:   fmt.Sprintf("%s_logger", clientName),
 		Output: os.Stdout,
 		Level:  hclog.Debug,
 	})
-	clientConfig.Logger = logger
-	client := plugin.NewClient(&clientConfig)
-	return &Wrapper{
-		Client: client,
-		Log:    logging.Default(),
-	}, nil
+	l := logging.Default()
```

Comment:
It would help me read logs if we added at least the plugin name as a field to the logger.  Similarly, we should output a log here giving _all_ the properties of the plugin: name, protocol, executable location, and that it's being used.